File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/feature/onboarding/flow/CableQrSheet.kt
package app.getvela.wallet.feature.onboarding.flow

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import app.getvela.wallet.core.designsystem.theme.VelaTheme
import app.getvela.wallet.core.designsystem.tokens.VelaFontFamily
import app.getvela.wallet.core.designsystem.tokens.VelaFontWeight
import app.getvela.wallet.core.designsystem.tokens.VelaSpacing
import app.getvela.wallet.core.designsystem.tokens.VelaTextSize
import app.getvela.wallet.core.i18n.I18nKeys
import app.getvela.wallet.core.i18n.LocalVelaStrings
import uniffi.vela_core_uniffi.cableQrMatrix

/**
 * "Sign in with your phone": the caBLE QR the OTHER device scans. Shown while a
 * [KeyMethod.Hybrid] ceremony is finding and talking to that phone; cleared the
 * moment it connects or fails. The matrix comes from the core (`cableQrMatrix`,
 * the same encoder every platform draws with), so the shell owns only pixels.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CableQrSheet(payload: String) {
    val strings = LocalVelaStrings.current
    val colors = VelaTheme.colors
    val matrix = remember(payload) { cableQrMatrix(payload) }

    ModalBottomSheet(
        // Dismissing is allowed — it simply lets the scan window time out — but
        // there is nothing to answer here; the person acts on the other phone.
        onDismissRequest = {},
        containerColor = colors.bgRaised,
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = VelaSpacing.xl2)
                .padding(bottom = VelaSpacing.xl3),
            verticalArrangement = Arrangement.spacedBy(VelaSpacing.lg),
        ) {
            // Reusing the hybrid-method copy ("Phone or tablet" / "Scan a code
            // …on a nearby device") rather than minting two new corpus keys
            // across 15 locales for a first cut; a dedicated SCAN_TITLE/BODY is
            // a follow-up through the i18n gate.
            Text(
                text = strings.t(I18nKeys.Create.METHOD_HYBRID_TITLE),
                color = colors.fgBase,
                fontFamily = VelaFontFamily,
                fontWeight = VelaFontWeight.bold,
                fontSize = VelaTextSize.xl2,
            )
            Text(
                text = strings.t(I18nKeys.Create.METHOD_HYBRID_BODY),
                color = colors.fgMuted,
                fontFamily = VelaFontFamily,
                fontSize = VelaTextSize.base,
            )
            if (matrix != null) {
                val width = matrix.width.toInt()
                val dark = Color.Black
                val light = Color.White
                Canvas(
                    modifier = Modifier
                        .fillMaxWidth()
                        .aspectRatio(1f)
                        .padding(vertical = VelaSpacing.md),
                ) {
                    // A quiet zone keeps scanners happy; draw the light ground
                    // then only the dark modules.
                    val quiet = 2
                    val units = width + quiet * 2
                    val cell = size.minDimension / units
                    drawRect(color = light, size = Size(size.width, size.height))
                    for (row in 0 until width) {
                        for (col in 0 until width) {
                            if (matrix.modules[row * width + col]) {
                                drawRect(
                                    color = dark,
                                    topLeft = Offset((col + quiet) * cell, (row + quiet) * cell),
                                    size = Size(cell, cell),
                                )
                            }
                        }
                    }
                }
            }
        }
    }
}

File: app-ios/VelaWallet/VelaWallet/Features/Onboarding/CableQrSheet.swift
//
//  CableQrSheet.swift
//  VelaWallet
//
//  "Sign in with your phone": the caBLE QR the OTHER device scans. Shown while
//  a Hybrid ceremony is finding and talking to that phone; replaced by the
//  touch sheet the moment the phone connects, and cleared however the ceremony
//  ends. The matrix comes from the core (cableQrMatrix, the same encoder every
//  platform draws with), so this view owns only pixels.
//

import SwiftUI
import VelaCore

struct CableQrSheet: View {
    @Environment(\.theme) private var theme
    let loc: Loc
    let payload: String

    var body: some View {
        VStack(spacing: Tokens.Space.s16) {
            // Reusing the hybrid-method copy ("Phone or tablet" / "Scan a code
            // …on a nearby device") rather than minting new corpus keys across
            // 15 locales for a first cut; dedicated copy is an i18n-gate
            // follow-up.
            Text(loc.t(I18nKeys.Create.methodHybridTitle))
                .typeRole(Typography.title)
                .foregroundStyle(theme.fgBase)
                .multilineTextAlignment(.center)

            Text(loc.t(I18nKeys.Create.methodHybridBody))
                .typeRole(Typography.body)
                .foregroundStyle(theme.fgMuted)
                .multilineTextAlignment(.center)
                .fixedSize(horizontal: false, vertical: true)

            if let matrix = cableQrMatrix(text: payload) {
                QrView(matrix: matrix)
                    .aspectRatio(1, contentMode: .fit)
                    .frame(maxWidth: 260)
                    .padding(.top, Tokens.Space.s8)
            }
        }
        .frame(maxWidth: .infinity)
        .padding(.horizontal, Tokens.Layout.screenPaddingX)
        .padding(.vertical, Tokens.Space.s32)
        .presentationDetents([.large])
        .presentationDragIndicator(.hidden)
        .presentationBackground(theme.bgRaised)
        // Dismissing would strand a ceremony blocked on the scan; it times out
        // on its own instead, exactly like the touch sheet.
        .interactiveDismissDisabled(true)
    }
}

/// The module matrix as pixels: a light card with a quiet zone, dark modules
/// only. Always black-on-white regardless of theme — scanners want contrast,
/// not palette.
private struct QrView: View {
    let matrix: QrMatrix

    var body: some View {
        Canvas { context, size in
            let width = Int(matrix.width)
            let quiet = 2
            let units = CGFloat(width + quiet * 2)
            let cell = min(size.width, size.height) / units
            context.fill(
                Path(CGRect(origin: .zero, size: size)),
                with: .color(.white)
            )
            for row in 0..<width {
                for col in 0..<width {
                    guard matrix.modules[row * width + col] else { continue }
                    let rect = CGRect(
                        x: (CGFloat(col) + CGFloat(quiet)) * cell,
                        y: (CGFloat(row) + CGFloat(quiet)) * cell,
                        width: cell,
                        height: cell
                    )
                    context.fill(Path(rect), with: .color(.black))
                }
            }
        }
    }
}
